spring-kafka-docs/src/main/antora/modules/ROOT/pages/kafka/transactions.adoc

```diff
@@ -79,6 +79,9 @@ Instead, use a `KafkaTransactionManager` in the container to start the Kafka tra
 
 See xref:tips.adoc#ex-jdbc-sync[Examples of Kafka Transactions with Other Transaction Managers] for an example application that chains JDBC and Kafka transactions.
 
+IMPORTANT: Container transactions use xref:retrytopic.adoc[Non-Blocking Retries] will break transaction atomicity.
+When listener code threw exception, container transactions commit success and record sending to the retryable topic.
```

Comment:
Isn't that what expected from asynchronous, non-blocking retries?
To be able to send to the retry topic we indeed has to commit the current transaction.
I'm not fully sure how this non-blocking retires work internally, but apparently we don't start a new transaction, when we consume from the next retry topic.

It might work some way with a `@Transactional` on the `@KafkaListener`, but pay attention that `@EnableTransactionManagement` has to be activated.

Either way: I would not say in the docs `break`.
It sounds like a bug which is not supposed to be there.

It is just a feature which cannot be combined with transactions due to async nature.

Do you see things differently?

Thanks

Reply:
Totally agree with you, i'll re-test the `@Transactional` case.

Reply:
Test result is non-blocking retries cannot be combined with transactions.

---

`@Transactional` case as follows:
When `PROPAGATION_REQUIRED` will infinite retry because of container transactions rollback. (un-expected).
When `PROPAGATION_REQUIRED_NEW`
1. If listener code fail and container code success, listener code rollback, container code commit success. (expected)
2. If listener code fail and container code fail, listener code rollback, container code rollback. (expected)
3. If listener code success and container code fail, listener code commit success, container code rollback then infinite retry. (un-expected)
Current `KafkaTransactionManager` only support `PROPAGATION_REQUIRES_NEW` and `PROPAGATION_REQUIRED`, the above case tests only these two cases.

PROPAGATION_REQUIRED_NEW un-expected Report see https://github.com/Wzy19930507/spring-kafka/tree/report_pr_3072

Please see if there are any omissions, thanks.

---

I'll modify docs to **non-blocking retries cannot be combined with transactions** first and see if there's a way to implement it.

Reply:
@Wzy19930507 We might want to use present tense here: `When the listener code throws an exception, container transaction commit succeeds, and the record is sent to the retryable topic.`

Reply:
Thanks for the correction, fix it.